src/hotspot/share/gc/parallel/psParallelCompact.cpp

```diff
@@ -1705,58 +1704,51 @@ void PSParallelCompact::fill_range_in_dense_prefix(HeapWord* start, HeapWord* en
   } while (addr < end);
 }
 
-void PSParallelCompact::fill_dead_objs_in_dense_prefix(uint worker_id, uint num_workers) {
+void PSParallelCompact::fill_dead_objs_in_dense_prefix() {
   ParMarkBitMap* bitmap = mark_bitmap();
 
   HeapWord* const bottom = _space_info[old_space_id].space()->bottom();
   HeapWord* const prefix_end = dense_prefix(old_space_id);
 
-  if (bottom == prefix_end) {
-    return;
-  }
-
-  size_t bottom_region = _summary_data.addr_to_region_idx(bottom);
-  size_t prefix_end_region = _summary_data.addr_to_region_idx(prefix_end);
-
-  size_t start_region;
-  size_t end_region;
-  split_regions_for_worker(bottom_region, prefix_end_region,
-                           worker_id, num_workers,
-                           &start_region, &end_region);
+  const size_t region_size = ParallelCompactData::RegionSize;
 
-  if (start_region == end_region) {
-    return;
-  }
+  // Fill dead space in [start_addr, end_addr)
+  HeapWord* const start_addr = bottom;
+  HeapWord* const end_addr   = prefix_end;
 
-  HeapWord* const start_addr = _summary_data.region_to_addr(start_region);
-  HeapWord* const end_addr = _summary_data.region_to_addr(end_region);
+  for (HeapWord* cur_addr = start_addr; cur_addr < end_addr; /* empty */) {
+    RegionData* cur_region_ptr = _summary_data.addr_to_region_ptr(cur_addr);
+    if (cur_region_ptr->data_size() == region_size) {
```

Comment:
The next lines up to `cur_addr += cur_region_ptr->partial_obj_size()` could possibly be extracted to a common function? Then you could check again if `cur_region_ptr->data_size() == region_size` to decide if you should `continue`.

Reply:
I see; yes, they are (almost) identical. However, there is a `continue` in the middle, which makes it hard to extract them out to be a function and these operations are quite caller specific, so forcing them into a function can hinder the readability, IMO, i.e. requiring readers to read into the implementation to understand the flow.